### ISSUE TITLE
PackAsTool generate shim in Build instead of Pack

### ISSUE THREAD

**Issue by wli3** [dotnet/cli#9288]
PackagedShimOutputDirectory currently set to temporary folder under IntermediateOutputPath. The producer (code gen in this case) need to unpack, sign and repack.

This ask is to generate shim during publish instead of pack so there is no repack required

**Comment by wli3** [comment]
relate issue https://github.com/dotnet/sdk/pull/2239

I maybe able to just move this task execution from pack to publish. However, there are some package context needed like packageid, packageversion which is used in part of the embedded path. Maybe that's fine due to they are properties. But we do need some design on that interface.

**Comment by wli3** [comment]
@seancpeters @vijayrkn

**Comment by nguerrera** [comment]
It's not really under **Base**IntermediateOutputPath, but actually IntermediateOutputPath (no base), right? Base path is shared between configurations and writing there would break parallel packing of different configurations.

**Comment by wli3** [comment]
@nguerrera oops, corrected it in description. You are right, just IntermediateOutputPath. i typed it wrong in the comment.

**Comment by mikem8361** [comment]
I need this issue resolved to sign my symbol downloader extension with the roslyn repotoolset sign tool.

/cc: @jaredpar

**Comment by nguerrera** [comment]
And the issue description should note explicitly that the shims should be put in the publish output folder, not merely be generated during publish but still to intermediate directory

**Comment by seancpeters** [comment]
+ @mlorbetske

**Comment by wli3** [comment]
I read our email thread again, and I think I get more confused. I need to understand signing more. From my understanding, the original issue description will not solve the problem. What Nick described will. At the same time the later is trivial while the first one is not.

**Comment by mikem8361** [comment]
I pretty sure all we need for our signing tool is to either change PackagedShimOutputDirectory to point somewhere under the OutputPath ("bin") or allow PackagedShimOutputDirectory to be overridden by my project file.  I pretty sure whether this is done in Pack or Publish doesn't matter.  Since change the path is pretty simple maybe this should be in new issue.

**Comment by nguerrera** [comment]
@mikem8361 That would be sufficient to unblock you, but I want to discuss writing the shims during publish  (which goes to customizable PublishDir and is under Bin\ by default)  

This is a model that is easier for me to understand, and solves a superset of what is blocking you. I gather that @wli3 has strong reservations about it, so we'll have to discuss this some more on our side.

If we can't come up with a quick consensus on it, then we can do the simpler thing to unblock you sooner.

**Comment by mikem8361** [comment]
Sounds good to me.

**Comment by wli3** [comment]
I talked to @nguerrera we think we should generate shim during _build_. So _--no-build_ flow can work.

The downside is : 
1. the flag is called _PackAsTool_ not build as tool
2. shim is only useful for package, it is not relevant to _build_

Although it is just "odd", and nothing more get created if people doesn't set the shim RID. Also the whole gen shim feature is all about signing, might as well make it smooth.

**Comment by mikem8361** [comment]
So what time frame can something to fix this (make signing of the shims possible with our sign tool)?

Publishing symbol downloader extension is being delayed by this and I really need to get this out to our customers.

**Comment by wli3** [comment]
https://github.com/dotnet/sdk/pull/2447
